synapse/rest/media/v1/_base.py

```diff
@@ -197,3 +199,64 @@ def __init__(self, server_name, file_id, url_cache=False,
         self.thumbnail_height = thumbnail_height
         self.thumbnail_method = thumbnail_method
         self.thumbnail_type = thumbnail_type
+
+
+def get_filename_from_headers(headers):
+    """
+    Get the filename of the downloaded file by inspecting the
+    Content-Disposition HTTP header.
+
+    Args:
+        headers (twisted.web.http_headers.Headers): The HTTP
+        request headers.
+
+    Returns:
+        A Unicode string of the filename, or None.
+    """
+    content_disposition = headers.get(b"Content-Disposition", [b''])
+
+    # Decode the Content-Disposition header (cgi.parse_header requires
+    # unicode on Python 3, not bytes) the best we can.
+    try:
+        content_disposition = content_disposition[0].decode('utf8')
```

Comment:
surely we shouldn't decode the entire header as utf-8?

Reply:
Well, we can't decode it as ASCII (which is the original bug), which means that someone is putting UTF-8 literal bytes into the header (non Synapse implementations?). `cgi.parse` needs `str` on Python 3, as per the comment, we have to decode it as something.

Reply:
>  someone is putting UTF-8 literal bytes into the header (non Synapse implementations?)

I think the content disposition comes from clients for uploaded media?

> cgi.parse needs str

sounds like `cgi.parse_header` might be broken then :/. My reading of the RFC is that you ought to split out the params *before* you decode it, because different params might be encoded differently, and you don't know what encoding they are using.

OTOH: fair enough, I can see that this is going to work in the majority of cases, and reinventing `parse_header` is probably more hassle than its worth. If you want to take that approach, can you expand the comment to record some of this?

Reply:
oh wait you rewrote it already. ugh. Ignore the above; I'll re-review this after dinner.